core/src/main/java/org/infinispan/util/concurrent/locks/impl/InfinispanLock.java

```diff
@@ -342,14 +346,14 @@ public void addListener(LockListener listener) {
       @Override
       public void cancel(LockState state) {
          checkValidCancelState(state);
-         out:
          do {
             LockState currentState = lockState;
             switch (currentState) {
                case WAITING:
                   if (casState(LockState.WAITING, state)) {
+                     onCanceled(this);
```

Comment:
I was thinking about moving `onCancelled()` before `notifyListeners()`. But I guess you're right, because I was able to reproduce the test failure again :)